File: src-tauri/src/memory.rs
use rusqlite::{params, Connection, Result as SqlResult};
use once_cell::sync::OnceCell;
use std::sync::Mutex;
use std::path::PathBuf;

static DB: OnceCell<Mutex<Connection>> = OnceCell::new();

fn get_db_path() -> PathBuf {
    let appdata = std::env::var("APPDATA").unwrap_or_else(|_| ".".to_string());
    let dir = PathBuf::from(appdata).join("Yusra");
    std::fs::create_dir_all(&dir).ok();
    dir.join("memory.db")
}

fn get_conn() -> &'static Mutex<Connection> {
    DB.get_or_init(|| {
        let path = get_db_path();
        let conn = Connection::open(path).expect("Failed to open SQLite database");
        init_tables(&conn).expect("Failed to initialize tables");
        Mutex::new(conn)
    })
}

fn init_tables(conn: &Connection) -> SqlResult<()> {
    conn.execute_batch(
        "CREATE TABLE IF NOT EXISTS episodic_memory (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            prompt TEXT NOT NULL,
            response TEXT NOT NULL,
            timestamp DATETIME DEFAULT CURRENT_TIMESTAMP,
            links_json TEXT DEFAULT '[]'
        );
        CREATE TABLE IF NOT EXISTS entity_state (
            key TEXT PRIMARY KEY,
            value_json TEXT NOT NULL,
            updated_at DATETIME DEFAULT CURRENT_TIMESTAMP
        );
        CREATE TABLE IF NOT EXISTS downloaded_models (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            model_name TEXT NOT NULL UNIQUE,
            local_path TEXT NOT NULL,
            is_active BOOLEAN DEFAULT FALSE,
            size_bytes INTEGER NOT NULL
        );"
    )?;
    Ok(())
}

pub fn store_memory(prompt: String, response: String) -> bool {
    let conn = get_conn().lock().unwrap();
    conn.execute(
        "INSERT INTO episodic_memory (prompt, response) VALUES (?1, ?2)",
        params![prompt, response],
    ).is_ok()
}

pub fn search_memory(query: String) -> String {
    let conn = get_conn().lock().unwrap();
    let search_pattern = format!("%{}%", query);
    let mut stmt = match conn.prepare(
        "SELECT id, prompt, response, timestamp FROM episodic_memory WHERE prompt LIKE ?1 OR response LIKE ?1 ORDER BY timestamp DESC LIMIT 10"
    ) {
        Ok(s) => s,
        Err(_) => return "[]".to_string(),
    };
    let results: Vec<String> = stmt.query_map(params![search_pattern], |row| {
        let id: i64 = row.get(0)?;
        let prompt: String = row.get(1)?;
        let response: String = row.get(2)?;
        let timestamp: String = row.get(3)?;
        Ok(format!("{{\"id\":{},\"prompt\":\"{}\",\"response\":\"{}\",\"timestamp\":\"{}\"}}", id, prompt.replace('"', "\\\""), response.replace('"', "\\\""), timestamp))
    })
    .unwrap()
    .filter_map(|r| r.ok())
    .collect();
    format!("[{}]", results.join(","))
}

pub fn set_entity(key: String, value: String) -> bool {
    let conn = get_conn().lock().unwrap();
    conn.execute(
        "INSERT OR REPLACE INTO entity_state (key, value_json, updated_at) VALUES (?1, ?2, datetime('now'))",
        params![key, value],
    ).is_ok()
}

pub fn get_entity(key: String) -> String {
    let conn = get_conn().lock().unwrap();
    let mut stmt = match conn.prepare("SELECT value_json FROM entity_state WHERE key = ?1") {
        Ok(s) => s,
        Err(_) => return "".to_string(),
    };
    stmt.query_row(params![key], |row| row.get::<_, String>(0))
        .unwrap_or_default()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_store_and_search_memory() {
        let stored = store_memory("test prompt".to_string(), "test response".to_string());
        assert!(stored);
        let results = search_memory("test".to_string());
        assert!(results.contains("test prompt"));
    }

    #[test]
    fn test_entity_state() {
        set_entity("test_key".to_string(), "\"test_value\"".to_string());
        let val = get_entity("test_key".to_string());
        assert!(val.contains("test_value"));
    }
}

File: backend/memory.py
"""Hybrid memory — SQLite (episodic graph + KV entity state). Schema per DEV.md §6."""
import json
import sqlite3
import threading
from pathlib import Path
from typing import Any, Optional

_lock = threading.Lock()
_conn: Optional[sqlite3.Connection] = None

DATA_DIR = Path.home() / ".yusra"
DB_PATH = DATA_DIR / "yusra.db"
MODELS_DIR = DATA_DIR / "models"

SCHEMA = """
CREATE TABLE IF NOT EXISTS episodic_memory (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    prompt TEXT NOT NULL,
    response TEXT NOT NULL,
    timestamp DATETIME DEFAULT CURRENT_TIMESTAMP,
    embedding_blob BLOB,
    links_json TEXT DEFAULT '[]',
    session_id TEXT,
    risk_level TEXT,
    actions_taken TEXT DEFAULT '[]'
);
CREATE INDEX IF NOT EXISTS idx_episodic_timestamp ON episodic_memory(timestamp);
CREATE INDEX IF NOT EXISTS idx_episodic_session ON episodic_memory(session_id);

CREATE TABLE IF NOT EXISTS entity_state (
    key TEXT PRIMARY KEY,
    value_json TEXT NOT NULL,
    updated_at DATETIME DEFAULT CURRENT_TIMESTAMP
);

CREATE TABLE IF NOT EXISTS downloaded_models (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    model_name TEXT NOT NULL UNIQUE,
    local_path TEXT NOT NULL,
    size_bytes INTEGER NOT NULL,
    is_active BOOLEAN DEFAULT FALSE,
    downloaded_at DATETIME DEFAULT CURRENT_TIMESTAMP,
    last_used_at DATETIME
);

CREATE TABLE IF NOT EXISTS sessions (
    id TEXT PRIMARY KEY,
    started_at DATETIME DEFAULT CURRENT_TIMESTAMP,
    ended_at DATETIME,
    persona_used TEXT NOT NULL,
    total_messages INTEGER DEFAULT 0,
    total_actions INTEGER DEFAULT 0
);
"""

DEFAULTS = {
    "active_persona": '"yusra"',
    "singularity_traits": '{"formality":0.5,"verbosity":0.5,"warmth":0.5,"humor":0.5,"proactivity":0.5,"custom_traits":[]}',
    "custom_persona": '""',
    "inference_temperature": "0.4",
    "inference_top_p": "0.9",
    "inference_max_tokens": "1024",
    "terminal_shell": '"powershell"',
    "terminal_cwd": '"~"',
    "onboarding_complete": "false",
}


def init_db() -> None:
    global _conn
    DATA_DIR.mkdir(parents=True, exist_ok=True)
    MODELS_DIR.mkdir(parents=True, exist_ok=True)
    with _lock:
        _conn = sqlite3.connect(DB_PATH, check_same_thread=False)
        _conn.row_factory = sqlite3.Row
        _conn.execute("PRAGMA journal_mode=WAL")
        _conn.executescript(SCHEMA)
        for key, value in DEFAULTS.items():
            _conn.execute(
                "INSERT OR IGNORE INTO entity_state (key, value_json) VALUES (?, ?)",
                (key, value),
            )
        _conn.commit()


def store_memory(prompt: str, response: str, session_id: str, actions: list[dict]) -> int:
    with _lock:
        cur = _conn.execute(
            "INSERT INTO episodic_memory (prompt, response, session_id, actions_taken) VALUES (?, ?, ?, ?)",
            (prompt, response, session_id, json.dumps(actions)),
        )
        _conn.commit()
        return cur.lastrowid


def search_memories(query: str, limit: int = 20) -> list[dict]:
    pattern = f"%{query}%"
    with _lock:
        rows = _conn.execute(
            "SELECT id, prompt, response, timestamp, session_id FROM episodic_memory "
            "WHERE prompt LIKE ? OR response LIKE ? ORDER BY timestamp DESC LIMIT ?",
            (pattern, pattern, limit),
        ).fetchall()
    return [dict(r) for r in rows]


def get_state(key: str, default: Any = None) -> Any:
    with _lock:
        row = _conn.execute("SELECT value_json FROM entity_state WHERE key = ?", (key,)).fetchone()
    if row is None:
        return default
    try:
        return json.loads(row["value_json"])
    except json.JSONDecodeError:
        return row["value_json"]


def set_state(key: str, value: Any) -> None:
    with _lock:
        _conn.execute(
            "INSERT OR REPLACE INTO entity_state (key, value_json, updated_at) VALUES (?, ?, datetime('now'))",
            (key, json.dumps(value)),
        )
        _conn.commit()
